src/solvers/lowering/byte_operators.cpp

```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
-
-            byte_extract_exprt byte_extract_expr(
-              src.id() == ID_byte_update_little_endian
-                ? ID_byte_extract_little_endian
-                : ID_byte_extract_big_endian,
-                src.value(), stored_value_offset,
-              element_size < sub_size ? src.value().type() : subtype);
-
-            new_value=lower_byte_extract(byte_extract_expr, ns);
-          }
-
-          // Where does the value we just extracted align in this cell?
-          // This value might be negative, indicating only the most-significant
-          // bytes should be written, to the least-significant bytes of the
-          // target array cell.
-          exprt overwrite_offset;
-          if(is_first_cell)
-            overwrite_offset = mod_exprt(src.offset(), sub_size_expr);
-          else if(is_last_cell)
-          {
-            // This is intentionally negated; passing is_last_cell
-            // to flatten below leads to it being negated again later.
-            overwrite_offset = minus_exprt(
-              minus_exprt(cell_offset, src.offset()), stored_value_offset);
-          }
-          else
-            overwrite_offset=zero_offset;
-
-          byte_update_exprt byte_update_expr(
-            src.id(),
-            old_cell_value,
-            overwrite_offset,
-            new_value);
-
-          // Call recursively, the array is gone!
-          // The last parameter indicates that the
-          exprt flattened_byte_update_expr=
-            lower_byte_update(byte_update_expr, ns, is_last_cell);
-
-          with_exprt with_expr(
-            result, cell_index, flattened_byte_update_expr);
-
-          result=with_expr;
-        }
-
-        return simplify_expr(result, ns);
-      }
+  if(is_array)
+    return simplify_expr(
+      array_exprt(std::move(elements), to_array_type(src.type())), ns);
+  else
+    return simplify_expr(
+      vector_exprt(std::move(elements), to_vector_type(src.type())), ns);
+}
+
+static exprt lower_byte_update_struct(
+  const byte_update_exprt &src,
+  const struct_typet &struct_type,
+  const array_exprt &value_as_byte_array,
+  const optionalt<exprt> &non_const_update_bound,
+  const namespacet &ns)
+{
+  exprt::operandst elements;
+  elements.reserve(struct_type.components().size());
+
+  mp_integer member_offset_bits = 0;
+  for(const auto &comp : struct_type.components())
+  {
+    auto element_width = pointer_offset_bits(comp.type(), ns);
+
+    // the next member would be misaligned, abort
+    if(
+      !element_width.has_value() || *element_width == 0 ||
+      *element_width % 8 != 0)
+    {
+      throw non_byte_alignedt(struct_type, comp, *element_width);
+    }
+
+    member_exprt member(src.op(), comp.get_name(), comp.type());
+
+    // compute the update offset relative to this struct member - will be
+    // negative if we are already in the middle of the update or beyond it
+    exprt offset = simplify_expr(
```

Comment:
const?

Reply:
No, it is conditionally reset to 0 later on.